rfd/0120-cloud-agent-auto-upgrades-ux.md

```diff
@@ -0,0 +1,68 @@
+---
+authors: Xin Ding (xin@goteleport.com)
+state: draft
+---
+
+# RFD 120 - Cloud Agent Auto Upgrades UX
+
+## Required Approvals
+
+* Engineering: @jimbishopp && @r0mant
+* Product: @klizhentas
+* Security: @reedloden
+
+## What
+
+Teleport will add support for Agent Auto Upgrades in v13.0.0. This document
+details tactical UX changes to expose the Agent Auto Upgrade capabilities to
+Teleport Cloud Admins.
+
+### Terminology
+
+* (Teleport) Practitioner: People who are actively using Teleport in their
+  day-to-day.
+* (Teleport) Admins: A subset of Practitioners. Admins deploy, configure, and
+  otherwise set up Teleport.
+* (Teleport) End Users: A subset of Practitioners. End Users use Teleport to
+  access infrastructure resources.
+* (Teleport) Agent: The `teleport` process which can run one more more Teleport
+  services.
+* (Teleport) Cluster Alerts: Messages in the Teleport Web UI and appropriate
+  Teleport CLIs that alert Practitioners of relevant concerns.
+* Unaccompanied Agent: A Teleport Agent deployed without an accompanying auto upgrader
+  service.
+
+## Why
+
+Simply adding support for Agent Auto Upgrades doesn't help if Admins don't adopt
+it. We want all Teleport customers to enable Agent Auto Upgrades in order to
+reduce the significant workload associated with manually managing and upgrading
+a fleet of Teleport Agents.
+
+## How
+
+There are two key tactical UX change that will increase the adoption of Agent
+Auto Upgrades among Teleport Cloud Admins.
+
+1. Where possible, push Admins to deploy Teleport Agents via supported package
+   managers (i.e. `helm`, `apt`, and `yum`) as this will deploy the
+   `teleport-ent-cloud-updater` package alongside the `teleport` package. It's
+   important that we provide a method (potentially via an optional flag with a
+   sane default) for Admins to point the auto upgrader service to the right
+   server. Post installation, the script should also inform Admins about how
+   Agent Auto Upgrades works, with an accompanying docs link to more details.
+2. Notify Admins via Cluster Alerts when Unaccompanied Agents are detected. The
+   Cluster Alert should also present a docs link informing Admins how to
+   properly add an auto upgrader service.
```

Comment:
@xinding33 Can we just update our existing "your agents are outdated" alerts with a link to docs on how to enable auto-upgrades for Cloud customers?

Reply:
@hugoShaka You're right, we likely don't need to technically actually identify Unaccompanied Agents. Admins will be happy as long as we solve the following: 
1. Notify Admins when Agents are not running the latest version.
2. Help Admins identify exactly which Agent(s) is/are problematic.
3. Help Admins "fix" said Agent(s) once and for all by providing instructions on 1) how to upgrade the Agent and 2) how to add an Agent Auto Upgrader Service to the Agent.

Reply:
@r0mant IMO, that's not sufficient because it doesn't satisfy 2 and 3 of the comment above.

Reply:
@xinding33 we've started discussing implementation of this RFD and need some clarification on a few points:

> Help Admins identify exactly which Agent(s) is/are problematic.

Do we actually want users to be able to  list all agents that are older than a given version and/or aren't configured to use an upgrader, or do we just want the alert message to include some examples of offending agents to help users catch agents that fell through the cracks?

Actually being able to list offending agents is obviously nicer, but is also a much more complex ask since it requires us to either implement the logic for tracking and filtering on upgrader status for each service type (brittle and annoying to maintain), or to enable the per-agent "instance" heartbeat which we switched to disabled by default when we opted to use the more weakly coordinated upgrade system (known to cause scalability issues for some etcd-based deployments).  Both options are doable, but neither is really ideal.

Including a few hostnames in an alert is comparatively easy, since each auth server can directly observe information about the agents connected to it, and write them to a resource.  But that has the downside of only really being useful in the case where agents are generally OK, but one or two fell through the cracks.

> Help Admins "fix" said Agent(s) once and for all by providing instructions on 1) how to upgrade the Agent and 2) how to add an Agent Auto Upgrader Service to the Agent.

Are you thinking of something more than just pointing a user at the docs?  If so what?

Reply:
@fspmarshall 

> Do we actually want users to be able to list all agents that are older than a given version and/or aren't configured to use an upgrader, or do we just want the alert message to include some examples of offending agents to help users catch agents that fell through the cracks?

Yes, we want users to be able to list all agents that are out-of-date. That's where the real value is because then users can take that list and fix everything. I don't have opinions on how we should go about the implementation. 

> Are you thinking of something more than just pointing a user at the docs? If so what?

We definitely want to link to docs but if we can provide a one-liner that helps users fix the problem 90% of the time, that'd be an excellent addition.